### PR TITLE
ws: Support conversations in authentication

### PR DESCRIPTION
**Pull request by petervo**
Supoprt authentication methods that take several requests / responses to complete. Start by supporting PAM Conversations.

Still needs unit tests and SSH support

### CODE REVIEW

File: src/ws/cockpitauth.c [hunk 1]
```diff
@@ -128,6 +130,104 @@ cockpit_authenticated_free (gpointer data)
...
+    }
+
+  auth_data_complete_result (ad, error);
+}
+
+static AuthData *
+auth_data_ref (AuthData *data)
+{
+  g_return_val_if_fail (data != NULL, NULL);
+  g_atomic_int_inc (&data->refs);
```
Comment (stefwalter) [inline src/ws/cockpitauth.c]: Does this need to be thread safe? The use of the atomic functions seem to indicate that it does. However we mutate the structure elsewhere, so I would say "no". Should we just use ```++``` here instead?

File: src/ws/cockpitauth.c [hunk 2]
```diff
@@ -128,6 +130,104 @@ cockpit_authenticated_free (gpointer data)
...
+  return data;
+}
+
+static void
+auth_data_unref (gpointer data)
+{
+  AuthData *d = data;
+  g_return_if_fail (data != NULL);
+  if (g_atomic_int_dec_and_test (&d->refs))
```
Comment (stefwalter) [inline src/ws/cockpitauth.c]: Ditto above.

**Comment by stefwalter** [comment]
This needs documentation of the authentication protocol (including reasons for its format). I guess that would go into doc/authentication.md?

**Comment by petervo** [comment]
@stefwalter, this is ready for review.

File: pkg/kubernetes/standalone/src/cockpit-kube-auth/main.go [hunk 3]
```diff
@@ -45,18 +45,18 @@ func jsonError(err error) ([]byte, error) {
 	return json.Marshal(data)
 }
 
-func readData(file *os.File) ([]byte, error) {
+func readData(fd int) ([]byte, error) {
 	result := make([]byte, MAX_BUFFER)
-	n, err := file.Read(result)
-	if err != nil && err != io.EOF {
+	n, err := syscall.Read(fd, result)
```
Comment (stefwalter) [inline pkg/kubernetes/standalone/src/cockpit-kube-auth/main.go]: Why is this change necessary? It needs explanation in the authentication.md documentation above. If all read operations *must* be at least 65536 bytes in length, then that needs to be stated.

File: src/static/login.html [hunk 4]
```diff
@@ -256,48 +293,75 @@
...
+            function send_login_request(method, headers, is_conversation) {
                 id("login-button").setAttribute('disabled', "true");
                 id("login-spinner").style.display = 'block';
                 var xhr = new XMLHttpRequest();
                 xhr.open("GET", login_path, true);
-                xhr.setRequestHeader("Authorization", "Basic " + btoa(utf8(user + ":" + password)));
+
+                var k;
+                for (k in headers)
+                    xhr.setRequestHeader(k, headers[k]);
```
Comment (stefwalter) [inline src/static/login.html]: This needs documentation somewhere. Maybe a new file login.md?

File: src/static/login.html [hunk 5]
```diff
@@ -325,6 +391,20 @@
...
+                    "Authorization": "Basic " + btoa(utf8(user + ":" + password))
+                };
+                send_login_request("GET", headers, false);
+            }
+
+            function converse(id, msg) {
+                var headers = {
+                    "Authorization": "Login-Reply " + id + " " + btoa(utf8(msg))
+                };
+                send_login_request("GET", headers, true);
```
Comment (stefwalter) [inline src/static/login.html]: My first reaction was to be concerned about the use of "GET" here for what is a non-idempotent action. However the remainder of the HTTP authentication mechanisms (other than Basic) do this too.

File: src/ws/session.c [hunk 6]
```diff
@@ -198,12 +196,50 @@ write_auth_begin (int result_code)
...
+  int r;
+
+  fprintf (authf, "}\n");
+  fflush (authf);
+  fclose (authf);
+
+
+  for (;;)
+    {
+      r = write (AUTH_FD, auth_msg, auth_msg_size);
```
Comment (stefwalter) [inline src/ws/session.c]: Corner case if you send zero bytes write() won't send a packet. Worth an assert() and a comment? Probably not worth switching to send() or sendmsg().

File: src/ws/session.c [hunk 7]
```diff
@@ -198,12 +196,50 @@ write_auth_begin (int result_code)
 }
 
 static void
+write_auth_begin (void)
+{
```
Comment (stefwalter) [inline src/ws/session.c]: Perhaps ```assert (authf == NULL);```?

File: src/ws/session.c [hunk 8]
```diff
@@ -198,12 +196,50 @@ write_auth_begin (int result_code)
...
+  authf = open_memstream (&auth_msg, &auth_msg_size);
+  fprintf (authf, "{ ");
+}
+
+static void
 write_auth_end (void)
 {
-  fprintf (authf, " }\n");
+  int r;
+
```
Comment (stefwalter) [inline src/ws/session.c]: Perhaps ```assert (authf != NULL);```

File: src/ws/session.c [hunk 9]
```diff
@@ -327,30 +366,55 @@ pam_conv_func (int num_msg,
...
+        {
+            debug ("answered pam password prompt");
+            resp[i].resp = *password;
+            resp[i].resp_retcode = 0;
+            *password = NULL;
+        }
+      else if (msg[i]->msg_style == PAM_ERROR_MSG)
         {
-          if (*password == NULL)
+          if (asprintf (&err_msg, "%s ", msg[i]->msg) < 0)
```
Comment (stefwalter) [inline src/ws/session.c]: Why are we reallocating this? Would a const char suffice? If not, then a comment would be good.

File: src/ws/session.c [hunk 10]
```diff
@@ -327,30 +366,55 @@ pam_conv_func (int num_msg,
...
+          *password != NULL)
+        {
+            debug ("answered pam password prompt");
+            resp[i].resp = *password;
+            resp[i].resp_retcode = 0;
+            *password = NULL;
+        }
+      else if (msg[i]->msg_style == PAM_ERROR_MSG)
         {
-          if (*password == NULL)
```
Comment (stefwalter) [inline src/ws/session.c]: Should we warnx() here too? Same as in the PAM_TEXT_INFO block below?

File: src/ws/session.c [hunk 11]
```diff
@@ -327,30 +366,55 @@ pam_conv_func (int num_msg,
...
-              *password = NULL;
+              break;
             }
         }
-      else if (msg[i]->msg_style == PAM_ERROR_MSG ||
-               msg[i]->msg_style == PAM_TEXT_INFO)
+      else if (msg[i]->msg_style == PAM_TEXT_INFO)
         {
-          warnx ("pam: %s", msg[i]->msg);
+          if (asprintf (&txt_msg, "%s ", msg[i]->msg) < 0)
```
Comment (stefwalter) [inline src/ws/session.c]: Why reallocate?

File: src/ws/session.c [hunk 12]
```diff
@@ -327,30 +366,55 @@ pam_conv_func (int num_msg,
...
       else
         {
-          warnx ("pam asked us for an unsupported info: %s", msg[i]->msg);
-          success = 0;
+          debug ("prompt for more data");
+          write_auth_begin ();
+          if (txt_msg)
+            write_auth_string ("message", txt_msg);
+          if (err_msg)
+            write_auth_string ("error", err_msg);
```
Comment (stefwalter) [inline src/ws/session.c]: How do we convey PAM_PROMPT_ECHO_OFF and does login.html respect it?

File: src/ws/session.c (hunk 9 above)
Comment (stefwalter) [inline src/ws/session.c]: Memory leak if multiple things set err_msg. But see above.

File: src/ws/session.c [hunk 13]
```diff
@@ -327,30 +366,55 @@ pam_conv_func (int num_msg,
...
+              break;
             }
         }
-      else if (msg[i]->msg_style == PAM_ERROR_MSG ||
-               msg[i]->msg_style == PAM_TEXT_INFO)
+      else if (msg[i]->msg_style == PAM_TEXT_INFO)
         {
-          warnx ("pam: %s", msg[i]->msg);
+          if (asprintf (&txt_msg, "%s ", msg[i]->msg) < 0)
+            {
```
Comment (stefwalter) [inline src/ws/session.c]: Memory leak if multiple things set txt_msg. But see above.

File: src/ws/cockpitauthpipe.c [hunk 14]
```diff
@@ -0,0 +1,466 @@
...
+
+  if (self->sig_pipe_close > 0)
+    g_signal_handler_disconnect (self->pipe, self->sig_pipe_close);
+  self->sig_pipe_close = 0;
+
+  g_clear_object (&self->pipe);
+}
+
+int
+cockpit_auth_pipe_claim_fd (CockpitAuthPipe *self)
```
Comment (stefwalter) [inline src/ws/cockpitauthpipe.c]: 'steal' is the normal word to use here, instead of 'claim'

File: src/ws/cockpitauthpipe.c [hunk 15]
```diff
@@ -0,0 +1,466 @@
...
+                             "out-fd", pair[1],
+                             "name", self->logname,
+                             "read-size", MAX_AUTH_BUFFER,
+                             NULL);
+
+  self->sig_pipe_read = g_signal_connect (self->pipe,
+                                           "read",
+                                           G_CALLBACK (on_pipe_read),
+                                           self);
+  self->sig_pipe_close = g_signal_connect (g_object_ref (self->pipe),
```
Comment (stefwalter) [inline src/ws/cockpitauthpipe.c]: Why doesn't self own self->pipe? Why does it need to be refed here?

File: src/ws/cockpitauthpipe.c [hunk 16]
```diff
@@ -0,0 +1,466 @@
...
+{
+  int pair[2];
+
+  CockpitAuthPipe *self = COCKPIT_AUTH_PIPE (object);
+
+  G_OBJECT_CLASS (cockpit_auth_pipe_parent_class)->constructed (object);
+
+  if (socketpair (AF_UNIX, SOCK_SEQPACKET, 0, pair) < 0)
+    {
+      g_error ("%s: Couldn't create socket pair: %s",
```
Comment (stefwalter) [inline src/ws/cockpitauthpipe.c]: EMFILE or ENFILE seem like errors that shouldn't kill the process? Or did you already consider them?

File: doc/authentication.md [hunk 17]
```diff
@@ -59,13 +59,22 @@ The command will be called with two arguments
...
+than 65536 bytes.
 
 By default cockpit-ws will wait a maximum of 30 seconds to receive this response.
 The number of seconds to wait can be adjusted by adding a timeout parameter along
 side the auth schema configuration in your config file. The given value should be
 a number between 1 and 900.
 
+If more information is needed the command should response with a json object containing
+a ```prompt``` string. These message will be displayed to the user and the user will be
+prompted for a response. If the user does not respond within 30 seconds the command will be
```
Comment (stefwalter) [inline doc/authentication.md]: 30 seconds seems low. What happens to the user when the timeout occurs. Do we indicate the timeout on the login screen with a small circle pie collapsing?

File: src/static/login.html (hunk 5 above)
Comment (petervo) [inline src/static/login.html]: Yep, I felt weird about this too. Originally i did it with POST, but cockpit goes out of it's way to only allow GET requests with empty bodies. I was hesitant to remove that restriction . But if you think that's better I still have the code lying around somewhere.

File: src/ws/session.c (hunk 12 above)
Comment (petervo) [inline src/ws/session.c]: Right now I have login html showing a password input for all prompts. Do you think we need to switch between text and password inputs based on echo?

File: pkg/kubernetes/standalone/src/cockpit-kube-auth/main.go (hunk 3 above)
Comment (petervo) [inline pkg/kubernetes/standalone/src/cockpit-kube-auth/main.go]: Because the file read waits for a EOF that never comes. So we allocate a buffer for the max size, do one read call on the fd and then only return what was actually read. The read doesn't need to be 65536 bytes.

Comment (stefwalter) [inline pkg/kubernetes/standalone/src/cockpit-kube-auth/main.go]: Makes sense. But that's exactly the sort of thing that needs to go in a comment.

File: doc/authentication.md (hunk 17 above)
Comment (petervo) [inline doc/authentication.md]: What would you suggest as a sane default? 

I don't have any indication on the login page of how long you have. Not sure how easy the animation will be to setup without jquery but i can look into it.

Comment (stefwalter) [inline doc/authentication.md]: > What would you suggest as a sane default? 

How long does it take to look for your two factor key press the key and type the code. Seems like it could be much longer than 30 seconds. What about a global timeout for the entire authentication rather than per response?

> I don't have any indication on the login page of how long you have. Not sure how easy the animation will be to setup without jquery but i can look into it.

That can be a follow up. Pure CSS can work. I can probably help.

File: src/ws/cockpitauthpipe.c (hunk 14 above)
Comment (petervo) [inline src/ws/cockpitauthpipe.c]: I figured I keep it matching ```cockpit_ssh_agent_claim_fd```. But I can change it.

File: doc/authentication.md (hunk 17 above)
Comment (petervo) [inline doc/authentication.md]: > How long does it take to look for your two factor key press the key and
> type the code. Seems like it could be much longer than 30 seconds. What
> about a global timeout for the entire authentication rather than per
> response?

Originally this was a global timeout. But since we don't know how many 
steps total are needed a default global timeout would have to be quite 
long. So I thought this was a better approach. I'll bump to 60 seconds 
by default unless you think a global one is better.

File: src/ws/session.c (hunk 10 above)
Comment (petervo) [inline src/ws/session.c]: Since your comment is on a deleted line, I'm not sure what line you are referring to

Comment (stefwalter) [inline src/ws/session.c]: Just that we should warnx() when PAM_ERROR_MSG in addition to the other logic.

File: src/ws/session.c (hunk 9 above)
Comment (petervo) [inline src/ws/session.c]: redid this, the idea is to join all of these messages together and return them in a string.

File: src/ws/session.c (hunk 7 above)
Comment (petervo) [inline src/ws/session.c]: Added

File: src/ws/session.c (hunk 8 above)
Comment (petervo) [inline src/ws/session.c]: Added

File: src/ws/session.c (hunk 10 above)
Comment (petervo) [inline src/ws/session.c]: Added

File: src/static/login.html (hunk 4 above)
Comment (petervo) [inline src/static/login.html]: Added

File: src/static/login.html (hunk 5 above)
Comment (stefwalter) [inline src/static/login.html]: I think this is good. Stays true to the fact that it's authentication.

Can we make it 'X-Login-Reply'?

File: src/ws/session.c (hunk 12 above)
Comment (stefwalter) [inline src/ws/session.c]: Yup, I think so. That's an important part of supporting PAM's prompting.

File: src/ws/cockpitauthpipe.c (hunk 14 above)
Comment (stefwalter) [inline src/ws/cockpitauthpipe.c]: Yeah, lets change it for ease of readability. This stuff is complex enough as it is.

File: doc/login.md [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+name of the user that was just logged in. Additional fields may be present
+
+Other http codes are considered errors. Generally these are 401 or 403 http codes
+and should contain a json body with the following fields:
+
+ * error: Values of ```authentication-failed```, ```authentication-unavailable``` or ```permission-denied``` are translated to the appropriate cockpit error codes. Any other values are treated as generic errors
+ * message: Optional text with more details about the error.
+
+In some authentication setups addtional steps are required. When this happens
+the json object will have a prompt and a id field.
```
Comment (stefwalter) [inline doc/login.md]: Have you considered sending back the challenge in headers? Symmetric to the reply? Willing to discuss this. But I'd like to make sure we're getting this right.

**Comment by stefwalter** [comment]
Is the idea that this pull request handles expired passwords? Because if so:

```
sudo chage -d 0 stef
```

... produces this message on the login screen:  "Authentication failed: pam-error: Authentication token is no longer valid; new one required"

**Comment by stefwalter** [comment]
https://youtu.be/gXsOt6zgBs4

**Comment by petervo** [comment]
It handles it with --local-ssh. For it to work with pam, there's more work to do. But i figured that can be a separate PR

File: doc/login.md (hunk 18 above)
Comment (petervo) [inline doc/login.md]: I did yes, but with headers we are then restricted to 8192 bytes. I wasn't sure that would always be enough.

Comment (petervo) [inline doc/login.md]: Esp since we have to base64 encode the response which will inflate it further

Comment (stefwalter) [inline doc/login.md]: How about we include the id in the header, and further prompt data in the body?

File: src/ws/session.c (hunk 12 above)
Comment (petervo) [inline src/ws/session.c]: Added

**Comment by petervo** [comment]
Reworked to use a WWW-Authenticate challenge header. The id and the prompt are part of the header. Any additional messages and the echo setting are part of the json body.
